bokeh/models/mappers.py

```diff
@@ -63,3 +65,229 @@ class LinearColorMapper(ColorMapper):
     def __init__(self, palette=None, **kwargs):
         if palette is not None: kwargs['palette'] = palette
         super(LinearColorMapper, self).__init__(**kwargs)
+
+class SegmentedColorMapper(ColorMapper):
+    """ Map numbers in a range [*low*, *high*] linearly into a
+    sequence of colors which are defined by a seires of breaks
+    correlated to different colors.
+
+    For example, if the breaks are specified 0, 0.5, 1 and the
+    colors are defined as 'red', 'white', 'green', values 
+    between 0 and 0.5 will smoothy vary between red and white
+    while values between 0.5 and 1 will vary between white and
+    green.
+    """
+
+    palette = Seq(Color, help="""
+    A sequence of colors to use as the target palette for mapping.
+
+    This property can also be set as a ``String``, to the name of
+    any of the palettes shown in :ref:`bokeh_dot_palettes`.
+    """).accepts(Enum(Palette), lambda pal: getattr(palettes, pal))
+
+    segments = Seq(Float, help="""
+    A listing of values which are directly tied to the 
+    colors defined in the palette.  The list must have
+    at least two elements.  If there are exactly two
+    elements, then the resulting colormap will be
+    interpolated between these two values and across
+    all of the colors defined in the palette.  If there
+    are more than two colors, the length must match the
+    number of colors in the palette.  In this configuration
+    each segment value is tied to the palette color of the
+    same index.  This argument is optional.  If it is
+    not provided, the min and max of the data will be used.
+    """)
+
+    alpha = Seq(Float, help="""
+    A python array, the same length as the dictionary with the
+    desired alpha values
+    """
+    )
+
+    interpolationMethod = String(default='linear', help="""
+    A string which describes the method of interpolation.  The default, 'linear',
+    will result in a gradual transition from one color to the next.  Other
+    options are 'step' which will result in a single color for all values 
+    which fall between the segment markers (an effective floor function) 
+    """)
```

Comment:
I do think we should wait on this. The reason is that past experience has (repeatedly) shown that "string configuration on the python side to implement policy on the JS side" is a choice that will be regretted later. The way this should be handled is to make proper models to represent different transforms, and configure the interpolation method by configuring a transform model. Something like:
```
from bokeh.models.transforms import Step, Linear, Log

color_mapper.interpolation = Step()
```
One reason to do this is because it making the segmented color mapper to be simpler by allowing it to delegate to other models. It is important to do with some deliberation because there are other places these transforms could be used as well.